### PR TITLE
Proves 51

### PR DESCRIPTION
**Pull request by bri-gaia**

### CODE REVIEW

File: app/lib/Plugins/SearchEngine/Elastic8.php [hunk 1]
```diff
@@ -31,63 +31,67 @@
...
+	static private array $record_cache = [];
 
 	protected $elasticsearch_index_name = '';
-	protected $elasticsearch_base_url = '';
+	protected string $elasticsearch_base_url = '';
 
-	protected $version = 8;
+	protected int $version = 8;
 
 	# -------------------------------------------------------
```
Comment (bri-gaia) [inline app/lib/Plugins/SearchEngine/Elastic8.php]: Should we make these dividers consistent? remove from here or add in other files?

File: app/lib/Plugins/SearchEngine/Elastic8ConfigurationSettings.php [hunk 2]
```diff
@@ -60,64 +60,68 @@ class Elastic8ConfigurationSettings extends ASearchConfigurationSettings {
...
+			= _t("CollectiveAccess uses only a single index in an ElasticSearch setup. The name of that index can be set in the CollectiveAccess configuration.");
 		# ------------------------------------------------
-		$this->setting_hints[ __CA_ELASTICSEARCH_SETTING_RUNNING__ ]
-			= _t( "Install and start the ElasticSearch service. If it is already running, check your CollectiveAccess configuration (the ElasticSearch URL and index name in particular).",
-			$this->webserver_user );
-		$this->setting_hints[ __CA_ELASTICSEARCH_SETTING_INDEX_EXISTS__ ]
-			= _t( "If the service is running and can be accessed by CollectiveAccess but the index is missing, let CollectiveAccess generate a fresh index and create the related indexing mappings. There is a tool in support/utils." );
+		$this->setting_hints[__CA_ELASTICSEARCH_SETTING_RUNNING__]
+			= _t("Install and start the ElasticSearch service. If it is already running, check your CollectiveAccess configuration (the ElasticSearch URL and index name in particular).",
+			$this->webserver_user);
```
Comment (bri-gaia) [inline app/lib/Plugins/SearchEngine/Elastic8ConfigurationSettings.php:95]: `$this->webserver_user` is undefined. Should this be removed? do we know what this is expected to be?

File: app/lib/Plugins/SearchEngine/Elastic8.php (hunk 1 above)
Comment (kehh) [reply app/lib/Plugins/SearchEngine/Elastic8.php]: Remove them all!

File: app/lib/Plugins/SearchEngine/Elastic8.php [hunk 3]
```diff
@@ -31,63 +31,67 @@
...
-	static private $doc_content_buffer = [];
-	static private $update_content_buffer = [];
-	static private $delete_buffer = [];
-	static private $record_cache = [];
+	static private array $doc_content_buffer = [];
+	static private array $update_content_buffer = [];
+	static private array $delete_buffer = [];
+	static private array $record_cache = [];
 
 	protected $elasticsearch_index_name = '';
```
Comment (kehh) [inline app/lib/Plugins/SearchEngine/Elastic8.php]:
```suggestion
	protected string $elasticsearch_index_name = '';
```

File: app/lib/Plugins/SearchEngine/Elastic8.php [hunk 4]
```diff
@@ -163,74 +163,78 @@ public function refreshMapping( $force = false ) {
...
-	 *
 	 * @param int $subject_tablenum
 	 * @param array $subject_row_ids
 	 * @param int $content_tablenum
 	 * @param string $content_fieldnum
+	 * @param int $content_container_id
 	 * @param int $content_row_id
 	 * @param string $content
-	 * @param array $options
+	 * @param null $options
```
Comment (kehh) [inline app/lib/Plugins/SearchEngine/Elastic8.php]:
```suggestion
	 * @param null|array $options
```

File: app/lib/Plugins/SearchEngine/Elastic8.php [hunk 5]
```diff
@@ -163,74 +163,78 @@ public function refreshMapping( $force = false ) {
...
+	 * @throws ClientResponseException
+	 * @throws MissingParameterException
+	 * @throws ServerResponseException
+	 * @throws Exception
 	 */
 	public function updateIndexingInPlace(
-		$subject_tablenum, $subject_row_ids, $content_tablenum, $content_fieldnum, $content_container_id,
-		$content_row_id, $content, $options = null
+		int $subject_tablenum, array $subject_row_ids, int $content_tablenum, string $content_fieldnum,
+		int $content_container_id, int $content_row_id, string $content, $options = null
```
Comment (kehh) [inline app/lib/Plugins/SearchEngine/Elastic8.php]: can we make $options a nullable array?

File: app/lib/Plugins/SearchEngine/Elastic8.php [hunk 6]
```diff
@@ -322,30 +327,46 @@ public function init() {
...
 	 *
 	 * @return bool
+	 * @throws AuthenticationException
+	 * @throws ClientResponseException
+	 * @throws MissingParameterException
+	 * @throws ServerResponseException
+	 * @throws Exception
 	 */
-	public function truncateIndex( $table_num = null) {
+	public function truncateIndex(int $table_num = null): bool {
```
Comment (kehh) [inline app/lib/Plugins/SearchEngine/Elastic8.php]: Do we need this?
```suggestion
	public function truncateIndex(?int $table_num = null): bool {
```

File: app/lib/Plugins/SearchEngine/Elastic8.php [hunk 7]
```diff
@@ -515,72 +540,80 @@ public function commitRowIndexing() {
...
+	 *
+	 * @throws ApplicationException
+	 * @throws AuthenticationException
+	 * @throws ClientResponseException
+	 * @throws MissingParameterException
+	 * @throws ServerResponseException
+	 * @throws Exception
 	 */
 	public function removeRowIndexing(
 		int $subject_tablenum, int $subject_row_id, ?int $field_tablenum = null, $field_nums = null,
```
Comment (kehh) [inline app/lib/Plugins/SearchEngine/Elastic8.php]:
```suggestion
		int $subject_tablenum, int $subject_row_id, ?int $field_tablenum = null, ?array $field_nums = null,
```

File: app/lib/Plugins/SearchEngine/Elastic8.php (hunk 6 above)
Comment (bri-gaia) [reply app/lib/Plugins/SearchEngine/Elastic8.php]: would make sense to me?

File: app/lib/Plugins/SearchEngine/Elastic8Result.php [hunk 8]
```diff
@@ -30,53 +30,53 @@
...
 	# -------------------------------------------------------
-	public function getHits() {
+	public function getHits(): array {
 		return $this->hits;
 	}
 
 	# -------------------------------------------------------
-	public function numHits() {
-		return is_array( $this->hits ) ? sizeof( $this->hits ) : 0;
+	public function numHits(): ?int {
```
Comment (kehh) [inline app/lib/Plugins/SearchEngine/Elastic8Result.php]:
```suggestion
	public function numHits(): int {
```

File: app/lib/Plugins/SearchEngine/Elastic8.php (hunk 7 above)
Comment (bri-gaia) [reply app/lib/Plugins/SearchEngine/Elastic8.php]: this is currently `int|null|array` in the PHPDoc comment, happy to go with your suggestion and we can switch it back if we find it is needed!

**Comment by bri-gaia** [comment]
@kehh This has become a bit of a mess of commits, we can't really squash these on merge, but I could go through and rebase the changes before we merge in? (once we are happy with it)

File: app/lib/Plugins/SearchEngine/Elastic8ConfigurationSettings.php (hunk 2 above)
Comment (bri-gaia) [reply app/lib/Plugins/SearchEngine/Elastic8ConfigurationSettings.php:95]: "nuke it"

File: app/lib/Plugins/SearchEngine/Elastic8.php (hunk 7 above)
Comment (bri-gaia) [reply app/lib/Plugins/SearchEngine/Elastic8.php]: Updating the signature is incompatible with the base class, have reverted this change, but updated the PHPDoc tag.

**Pull request merged**